include/abstract_data_store.h

```diff
@@ -78,19 +80,31 @@ template <typename data_t> class AbstractDataStore
     // num_points) to zero
     virtual void copy_vectors(const location_t from_loc, const location_t to_loc, const location_t num_points) = 0;
 
-    // metric specific operations
-
+    // Some datastores like PQ stores need a preprocessing step before querying.
+    // Optionally, a scratch object can be passed in to avoid memory allocations
+    // Default implementation does nothing.
+    // REFACTOR TODO: Currently, we take an aligned_query as parameter, but this
+    // should change and this function should do the necessary alignment.
+    virtual void preprocess_query(const data_t *aligned_query, AbstractScratch<data_t> *query_scratch = nullptr) const;
+    // distance functions.
     virtual float get_distance(const data_t *query, const location_t loc) const = 0;
     virtual void get_distance(const data_t *query, const location_t *locations, const uint32_t location_count,
-                              float *distances) const = 0;
+                              float *distances, AbstractScratch<data_t> *scratch_space = nullptr) const = 0;
+    // Specific overload for index.cpp.
+    // REFACTOR TODO: Check if the default implementation is sufficient for most cases.
+    virtual void get_distance(const data_t *preprocessed_query, const std::vector<location_t> &ids,
+                              std::vector<float> &distances, AbstractScratch<data_t> *scratch_space) const = 0;
     virtual float get_distance(const location_t loc1, const location_t loc2) const = 0;
 
     // stats of the data stored in store
     // Returns the point in the dataset that is closest to the mean of all points
     // in the dataset
     virtual location_t calculate_medoid() const = 0;
 
-    virtual Distance<data_t> *get_dist_fn() = 0;
+    //REFACTOR PQ TODO: Each data store knows about its distance function, so this is 
+    //redundant. However, we don't have an OptmizedDataStore yet, and to preserve code 
+    //compability, we are exposing this function. 
+    virtual Distance<data_t>* get_dist_fn() const = 0;
```

Comment:
The distance in index is used to call preprocess_query() in diatnace function. we can add a method preprocess_query() here and that in turn will call it from distance. that way we do not need to expose distance.

Reply:
The only place we need it still is for fast_l2/optimized_distance metric.